README

= fuc

fuckin' useful crawler

== Description

simple cralwer

== Installation

=== Archive Installation

	rake install

=== Gem Installation

	gem install fuc


== Features/Problems


== Synopsis


== Copyright

Author::    koyachi <rtk2106@gmail.com>
Copyright:: Copyright (c) 2008 koyachi
License::   


ChangeLog
== 0.0.1 / 2008-11-25

* initial release



test/fuc_test.rb
require File.dirname(__FILE__) + '/test_helper.rb'

require "test/unit"
class FucTest < Test::Unit::TestCase
end


test/test_helper.rb
require 'test/unit'
require File.dirname(__FILE__) + '/../lib/fuc'



lib/fuc/model.rb
# -*- coding: utf-8 -*-
require 'rubygems'
require 'dm-core'
require 'dm-types'

DataMapper.setup(:default, "sqlite3:///#{Dir.pwd}/fuc.db")

def setup_model
  DataMapper.auto_migrate!
end

class Entry
  include DataMapper::Resource
  include DataMapper::Types

  property :id, Serial
  property :url, String
  property :via_url, String
  property :summary, Text
  property :body, Object #=>URLCacheのresponseみたいな使い方,
  property :checked, Boolean, :default => false
  property :created_at, DateTime, :default => Proc.new {|r,p| DateTime.now}

  has n, :tags, :through => Resource

  def should_crawl?
    !self.checked || self.created_at.to_time < DateTime.now.to_time - 10 * 60
  end
end

class Tag
  include DataMapper::Resource
  include DataMapper::Types
  
  property :id, Serial
  property :title, String

  has n, :entries, :through => Resource
  has n, :bundles, :through => Resource
end

# view, queue(host)
class Bundle
  include DataMapper::Resource
  include DataMapper::Types

  property :id, Serial
  property :title, String
  
  has n, :tags, :through => Resource
end

__END__

# 使用例

# 必要なbundle各自作ってそこから取り出すようにする
# push_urlにtagとbundle指定できるようにしないと。
class YourProcesser < Fuc::Processer
  include Fuc::Storage

  def match(url_entry)
    # 例えば[youtube|vimeo]なら
    nil
  end

  def process
    # viewバンドルのvideoタグつけてpush_url
  end
end

# 参照側
require 'fuc-processer-storage'
class Foo
  include Fuc::Storage

  def process_entry
    entry = Entry.find(0)
  end

  # 条件(type)指定してentry取得したいのはどんなときか?
  # - entry数が多くてselect likeでの抽出に時間がかかりそうな場合
  # - 
  def process_entry_by_some_type(type)
    type = Type.find(title => 'type_a')
    type.entries.each do |entry|
      
    end
  end
end


lib/fuc/queue.rb
# -*- coding: utf-8 -*-
module Fuc
  class URLQueue
    def initialize
      @current_slot = 0
      @queue = Bundle.first_or_create(:title => 'queue')
      _make_slotid_list
    end

    def _make_slotid_list
      @slot_ids = @queue.tags.map {|t| t.id}
      p @slot_ids
    end

    def _add_slotid(slot)
      @slot_ids.push(slot.id)
    end

    def push(url_entry)
      host = URI.split(url_entry.url.to_s)[2]
      slot = @queue.tags.first(:title => host)
      unless slot then
        p "create Tag #{host}"
        slot = Tag.create(:title => host)
        @queue.tags << slot
        @queue.save
        _add_slotid(slot)
      end
      slot.entries << url_entry
      slot.save
    end

    def _pop_one
      Entry.first(:checked => false)
    end

    # FIXME: 優先度低め Fuc.runのループと合わせて見直し
    def _pop_one_auto
      url_entry = nil
      nil_count = 0
      while url_entry == nil do
        p "@crrent_slot = #{@current_slot}"
        host_name = @queue.tags.get(@slot_ids[@current_slot])
        p host_name
        host = @queue.tags.first(:title => host_name.title)
        url_entry = host.entries.first(:checked => false)
        if url_entry == nil then
          host.destroy
          host.save
          nil_count = nil_count + 1
          if nil_count == @slot_ids.length then
            break
          end
          next
        end
        @current_slot = (@current_slot < @slot_ids.length - 1) ? @current_slot + 1 : 0
      end
      url_entry
    end

    # host毎にプロセス分けて並列処理するようなときはhostsとあわせてこっち使う
    def _pop_by_host(host)
      url_entry = @queue.tags(host).first
      url_entry
    end

    def pop(*args)
      case args.length
      when 0
        _pop_one
      when 1
        _pop_by_host(args[0])
      end
    end

    def hosts
      @queue.tags.map {|host| host.title}
    end
  end
end



lib/fuc/processer.rb
# -*- coding: utf-8 -*-
module Fuc
  class << self
    cattr_accessor :processers
    self.processers = []

    def register_process(user_class=Processer, force=false, type='normal')
      def _init(cls, force, type)
        return if !force && %w[Fuc::Processer::NotifyURLsVIA].any?{|c| c == cls.name}
        self.processers << cls.new(self)
      end

      if user_class.instance_of? Array then
        user_class.each do |cls|
          _init(cls, force, type)
        end
      else
        _init(user_class, force, type)
      end
    end
  end

  class Processer
    NAME = 'Processer'
    DESCRIPTION = 'generic processer'

    def initialize(crawler)
      setup()
      @crawler = crawler
    end

    def setup
    end

    def match(url_entry)
      false
    end

    def process(url_entry)
    end

    def self.inherited(subclass)
      Fuc.register_process subclass
    end

    def log(msg)
    end


    def push_urls(*url_entry)
      @crawler.url_queue.push(*url_entry)
    end

    def push_url(url, via_url, summary, tag='text')
      print "%-10s %s\n" % ["push", url]
      # tag, bundle
      # FIXME: Bundle対応
      tag = Tag.first_or_create(:title => tag)
      entry = Entry.new(:url => url,
                        :via_url => via_url,
                        :summary => summary)
      tag.entries << entry
      tag.save
      entry.save
      @crawler.url_queue.push(entry)
    end

    # FIXME ダサッ
    def download(url, savedir='', type='blob')
      ext = (type != 'blob') ? ".#{type}" : File.extname(url) or ''
      filename = ((savedir == '') ? Fuc.work_dir : savedir) + "/#{Digest::MD5.hexdigest(url)}#{ext}"
      print "  download #{url} to #{filename}\n"
      content = open(url).read
      File.open(filename, "w") {|f|
        f.write content
      }
      [filename, content]
    end

    def nokogiri(content)
      Nokogiri::HTML(content)
    end
  end

  class DownloadProcesser < Processer
    NAME = 'DownloadProcesser'
    DESCRIPTION = 'download somethin'

    def self.inherited(subclass)
      Fuc.register_process(subclass, false, 'downloader')
    end
  end

  class ContentExtractProcesser < Processer
    NAME = 'ContentExtractProcesser'
    DESCRIPTION = 'extract html content'

    def match(url_entry)
      url_entry.tags.first(:title => 'cotent_extracter')
    end

    def process(url_entry)
      # extract main content
    end

    def self.inherited(subclass)
      Fuc.register_process(subclass, false, 'content_extracter')
    end
  end
end


lib/fuc/processer/downloader.rb
class Fuc::Processer
  module Downloader
    def convert_url(url)
      url
    end

    def process(url_entry)
      download(convert_url(url_entry.url))
      url_entry.checked = true
      url_entry.save
    end
  end

  class ImageDownloader < Fuc::Processer
    NAME = 'ImageDownloader'
    DESCRIPTION = 'download image files url end with [jpg|jpeg|png|gif]'

    include Fuc::Processer::Downloader

    def match(url_entry)
      url_entry.url =~ %r!.*.[jpg|jpeg|png|gif]$!i
    end
  end

  class MusicDownloader < Fuc::Processer
    NAME = 'MusicDownloader'
    DESCRIPTION = 'download music files url end with [mp3]'

    include Fuc::Processer::Downloader

    def match(url_entry)
      url_entry.url =~ %r!.*.mp3!
    end
  end

#  class VideoDownloader < Fuc::Processer
#    include Fuc::Processer::Downloader
#
#    def match(url_entry)
#    end
#  end
end


lib/fuc.rb
#!/usr/bin/env ruby
# -*- coding: utf-8 -*-
#
# fuckin' useful crawler
# 2008-11-25 t.koyachi
require 'rubygems'
require 'open-uri'
require 'nokogiri'
require 'extlib'
require 'net/http'
require 'uri'
require 'tmpdir'
require 'digest/md5'

dir = Pathname(__FILE__).dirname.expand_path / 'fuc'
require dir / 'model'
require dir / 'queue'
require dir / 'processer'
require dir / 'processer' / 'notify_urls_via'
require dir / 'processer' / 'downloader'
require dir / 'processer' / 'webservices'

module Fuc
  VERSION = '0.0.2'

  class << self
    cattr_accessor :work_dir
    self.work_dir = Dir.tmpdir + '/fuc'

    def setup_workdir(dir='')
      self.workdir = dir if dir != ''
      Dir.mkdir(self.work_dir) unless File.directory? self.work_dir
    end

    cattr_accessor :url_queue

    def run(url)
      setup_workdir
      # FIXME
      setup_model
      self.url_queue = URLQueue.new

      if url.instance_of? String then
        self.url_queue.push(Entry.new(:url => url))
      elsif url.instance_of? Array then
        url.map {|u|
          self.url_queue.push(Entry.new(:url => url))
        }
      end
      
      while url_entry = self.url_queue.pop() do
        print "%-10s %s\n" % ["pop", url_entry.url]
        now = DateTime.now.to_time
        # FIXME: 直値 -> Config
        expire = (url_entry.via_url.nil?) ? now : now - 10 * 60
        # 前回チェックから一定時間が経っていなければスキップ
        unless url_entry.should_crawl? then
          print "skip\n"
          next
        end

        response = nil
        self.processers.each do |p|
          next unless p.match(url_entry)

          print "%-10s %s\n" % ["process", p.class.name]
          uri_parts = URI.split(url_entry.url)
          host, port, path = uri_parts[2], uri_parts[3],
                             (uri_parts[5] == '') ? '/index.html' : uri_parts[5]
          begin
            response = url_entry.body
            if response.nil? then
              Net::HTTP.start(host, port) {|http|
                response = http.head(path)
                if response['content-type'] =~ %r!text/html! then
                  response = http.get(path).body
                else
                  response = nil
                end
              }
              url_entry.body = response
              url_entry.save
              sleep 1
            end
            result = p.process(url_entry)
          rescue Timeout::Error
            print "timeout\n"
          end  
          #break
        end
        url_entry.checked = true
        url_entry.save
        print "\n"
      end
    end

    def output_processer_log(user_class)
      def mixin(cls)
        cls.class_eval <<-MIXIN
          def log(msg)
            print msg + "\n"
          end
        MIXIN
      end
      if user_class.instance_of? Array then
        user_class.each do |cls|
          mixin(cls)
        end
      else
        mixin(user_class)
      end
    end
  end
end

